File: ml/metrics.py
"""Metrics for evaluation."""

from __future__ import annotations

import numpy as np


def accuracy_score(y_true: np.ndarray, y_pred: np.ndarray) -> float:
    return np.mean(y_true == y_pred)


def mean_squared_error(y_true: np.ndarray, y_pred: np.ndarray) -> float:
    return np.mean((y_true - y_pred) ** 2)


def r2_score(y_true: np.ndarray, y_pred: np.ndarray) -> float:
    ss_res = np.sum((y_true - y_pred) ** 2)
    ss_tot = np.sum((y_true - np.mean(y_true)) ** 2)
    return 1 - ss_res / ss_tot if ss_tot != 0 else 0.0


def confusion_matrix(y_true: np.ndarray, y_pred: np.ndarray) -> np.ndarray:
    classes = np.unique(np.concatenate([y_true, y_pred]))
    cm = np.zeros((len(classes), len(classes)), dtype=int)
    for t, p in zip(y_true.astype(int), y_pred.astype(int)):
        cm[t, p] += 1
    return cm

File: ml/metrics.ts
export function accuracy_score(y_true: number[], y_pred: number[]): number {
  let correct = 0;
  for (let i = 0; i < y_true.length; i++) if (y_true[i] === y_pred[i]) correct++;
  return correct / y_true.length;
}

export function mean_squared_error(y_true: number[], y_pred: number[]): number {
  return y_true.reduce((s, y, i) => s + (y - y_pred[i]) ** 2, 0) / y_true.length;
}

export function r2_score(y_true: number[], y_pred: number[]): number {
  const mean = y_true.reduce((s, v) => s + v, 0) / y_true.length;
  const ss_res = y_true.reduce((s, y, i) => s + (y - y_pred[i]) ** 2, 0);
  const ss_tot = y_true.reduce((s, y) => s + (y - mean) ** 2, 0);
  return 1 - ss_res / ss_tot;
}

export function confusion_matrix(y_true: number[], y_pred: number[]): number[][] {
  const classes = [...new Set([...y_true, ...y_pred])].sort((a, b) => a - b);
  const n = classes.length;
  const cm = Array.from({ length: n }, () => new Array(n).fill(0));
  for (let i = 0; i < y_true.length; i++) {
    const ti = classes.indexOf(y_true[i]);
    const pi = classes.indexOf(y_pred[i]);
    if (ti >= 0 && pi >= 0) cm[ti][pi]++;
  }
  return cm;
}